store/iavl/iterator.go

```diff
@@ -0,0 +1,68 @@
+package iavl
+
+import (
+	"sync"
+
+	"github.com/iov-one/weave/store"
+)
+
+type lazyIterator struct {
+	data    store.Model
+	hasMore bool
+	read    chan store.Model
+	stop    chan struct{}
+	once    sync.Once
```

Comment:
I really hate the force concurrency here. Considered the not-so-great [`iavl` API](https://godoc.org/github.com/tendermint/iavl#ImmutableTree.IterateRange) I think you did a great job of fighting it and managing to expose a decent API after all :)

~I think certain implementation details can be improved. I will address them in separate comments to make it easier.~ I have removed all separate comments and summarized it all below.

I would like to propose a bit different approach. This is the easiest to explain with a code example. I hope you don't mind that I propose this.

```go
type iterator struct {                                                         
        next chan keyvalue                                                     
                                                                               
        stop func()                                                            
        done <- chan struct{}                                                  
}                                                                              
                                                                               
type keyvalue struct {                                                         
        key []byte                                                             
        val []byte                                                             
}                                                                              
                                                                               
func newIterator() *iterator {                                                 
        // Using context for synchronization is easier than writing this by hand.
        ctx, stop := context.WithCancel(context.Background())                  
                                                                               
        it := &iterator{                                                       
                next: make(chan keyvalue),                                     
                stop: stop,                                                    
                done: ctx.Done(),                                              
        }                                                                      
        return it                                                              
}                                                                              
                                                                               
// add is used as a callaback to iavl's IterateRange. This function should     
// return true if no more entries should be read. Returning false means "do not
// stop" and read another entry.                                               
// https://godoc.org/github.com/tendermint/iavl#ImmutableTree.IterateRange     
func (it *iterator) add(key, value []byte) (stop bool) {                       
        select {                                                               
        case it.next <- keyvalue{key: key, val: value}:                        
                return false                                                   
        case it.done:                                                          
                return true                                                    
        }                                                                      
}                                                                              
                                                                               
// Next returns the next key/value pair from the result set. If all the data   
// was returned by the itrator and there is no more content, ErrDone is        
// returned.                                                                   
func (it *iterator) Next() (key, value []byte, err error) {                    
        select {                                                               
        case pair := <-it.next:                                                
                return pair.key, pair.val, nil                                 
        case <-it.done:                                                        
                return nil, nil, ErrDone                                       
        }                                                                      
}                                                                              
                                                                               
                                                                               
// Stop finalize the iterator. It is the user responsibility to close the      
// itertor in order to free the resources. This method is safe to call any     
// number of times.                                                            
// XXX: This method is not called Close in order to avoid conflict with the    
// closer interface.                                                           
func (it *iterator) Stop() {                                                   
        it.stop()                                                              
}                                                                              
                                                                               
ErrDone = errors.Register(999, "done")                                         
                                                                               
// ----------------------------------                                          
                                                                               
                                                                               
// Iterator and ReverseIterator no longer have to return an error.             
//                                                                             
// They also do not close the iterator as this must be done by the user anyway.
// Not closing it with a defer makes it easier to synchronize.                 
func (a adapter) ReverseIterator(start, end []byte) store.Iterator {           
        it := newIterator()                                                    
        go a.tree.IterateRange(start, end, false, it.add)                      
        return it                                                              
}         
```

The main difference is
- I am using `context.Context` for synchronization because I am lazy :see_no_evil: 
- I have changed the `Next()` method to return key and value. This reduce the amount of methods and there is no need to the `ReadOneFromIterator` function anymore
- I am using `ErrDone` error to sygnalize that the iterator is finished
- `Iterator` and `ReverseIterator` methods no longer return an error
- I have renamed `Close` method to `Stop` in order to not conflict with the `io.Closer` interface

What do you think?

Reply:
Thank you for proposing this.

Context.WithCancel seems a bit heavy to me, when it really is just closing a channel (with your fix). Unless I accept a context as an argument (allowing user to cancel or pass timeout separately), I don't see the advantage.

Otherwise, they are good points, but require changing the Iterator API. This may be a good thing, but I think is out of scope of this PR. The main point of the current implementation is to allow some sort of for loops....

```go
func consumeIterator(itr weave.Iterator) ([]weave.Model, error) {
	defer itr.Close()

	var res []weave.Model
	for itr.Valid() {
		mod := weave.Model{
			Key:   itr.Key(),
			Value: itr.Value(),
		}
		res = append(res, mod)
		err := itr.Next()
		if err != nil {
			return nil, err
		}
	}
	return res, nil
}
```

I would like to see an example of such a for loop with the new API. (I will try below to compare)

Reply:
(This KVStore API was one of the first APIs in weave, so I am sure it is good to update it... I just want to reflect a bit more on that, and would like a separate PR, as this will change a number of points outside of the store package).

Reply:
This is my attempt (with type holders so it compiles).
Not so happy with it actually.... maybe you can do better?

```go
type Itr interface {
	Next() ([]byte, []byte, error)
	Stop()
}

var ErrDone = errors.Register(999, "done")

func consume(itr Itr) ([]weave.Model, error) {
	defer itr.Stop()

	var res []weave.Model
	var key, value []byte
	var err error
	for key, value, err = itr.Next(); err == nil; key, value, err = itr.Next() {
		mod := weave.Model{
			Key:   key,
			Value: value,
		}
		res = append(res, mod)
	}
	if err == ErrDone {
		err = nil
	}
	return res, nil
}
```